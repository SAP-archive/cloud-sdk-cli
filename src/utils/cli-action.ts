/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */
import cli from 'cli-ux';

export async function action(output: string, showSpinner: boolean, promise: Promise<any>) {
  if (showSpinner) {
    cli.action.start(output);
  } else {
    cli.log(`${output}...`);
  }
  await promise;
  if (showSpinner) {
    cli.action.stop();
  }
}
