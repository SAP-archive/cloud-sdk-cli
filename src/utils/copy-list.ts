/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */

export interface CopyDescriptor {
  sourcePath: string | URL;
  targetFolder: string;
  fileName: string;
}
