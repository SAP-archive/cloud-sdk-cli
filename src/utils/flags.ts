/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */

import Command from '@oclif/command';
import { OutputFlags } from '@oclif/parser';

type FlagsInput<CommandT extends typeof Command> = CommandT['flags'];

export type Flags<CommandT extends typeof Command> = OutputFlags<NonNullable<FlagsInput<CommandT>>>;
export type RequiredFlagsParam<CommandT extends typeof Command, ParamsT extends keyof Flags<CommandT>> = Pick<Flags<CommandT>, ParamsT>;
export type OptionalFlagsParam<CommandT extends typeof Command, ParamsT extends keyof Flags<CommandT>> = Partial<
  RequiredFlagsParam<CommandT, ParamsT>
>;
export type FlagsParam<
  CommandT extends typeof Command,
  RequiredParamsT extends keyof Flags<CommandT>,
  OptionalParamsT extends keyof Flags<CommandT>
> = RequiredFlagsParam<CommandT, RequiredParamsT> & OptionalFlagsParam<CommandT, OptionalParamsT>;
