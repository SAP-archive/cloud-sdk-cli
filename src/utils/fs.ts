/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { promisify } from 'util';
import * as fs from 'fs';
import * as rimraf from 'rimraf';

export const access = promisify(fs.access);
export const mkdir = promisify(fs.mkdir);
export const readdir = promisify(fs.readdir);
export const readFile = promisify(fs.readFile);
export const copyFile = promisify(fs.copyFile);
export const writeFile = promisify(fs.writeFile);
export const rm = promisify(rimraf);
