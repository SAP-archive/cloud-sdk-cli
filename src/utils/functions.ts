/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */

export interface Confirm {
  confirm: (message: string) => Promise<boolean>;
}

function confirm(message: string): Promise<boolean> {
  return Promise.resolve(true);
}

export const testFunctions = {
  confirm
};
