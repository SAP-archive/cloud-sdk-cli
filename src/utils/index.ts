/*!
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 */

export * from './copy-list';
export * from './generate-vdm-util';
export * from './git-ignore';
export * from './jest-config';
export * from './manifest-yaml';
export * from './message-formatter';
export * from './package-json';
export * from './scaffold';
export * from './templates';
export * from './usage-analytics';
export * from './warnings';
