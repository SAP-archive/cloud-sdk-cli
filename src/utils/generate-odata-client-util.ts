/*!
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 */

import { flags } from '@oclif/command';
import { AlphabetLowercase, AlphabetUppercase } from '@oclif/parser/lib/alphabet';
import { IBooleanFlag, IOptionFlag } from '@oclif/parser/lib/flags';
import * as path from 'path';
import { Options } from 'yargs';
import { GeneratorOptionsSDK } from './generator-options';

interface GeneratorOptionCli {
  projectDir: string;
}

type KeysToOptions = {
  [optionName in keyof GeneratorOptionCli]: Options;
};

export const generatorOptionCli: KeysToOptions = {
  projectDir: {
    default: '.',
    describe: 'Path to the folder in which the VDM should be created. The input and output dir are relative to this directory.',
    type: 'string'
  }
};

type AllOptions = GeneratorOptionsSDK & GeneratorOptionCli;

// OClif distinguishes between boolean and string flags. Split the keys to get proper typing
type FilterBooleanKeys<Base> = {
  [Key in keyof Base]: Base[Key] extends boolean ? Key : never;
};

type BoolArgKeys = NonNullable<FilterBooleanKeys<AllOptions>[keyof AllOptions]>;
export type BoolArgType = {
  [optionName in BoolArgKeys]: IBooleanFlag<boolean>;
};

type StringArgKeys = keyof Omit<AllOptions, BoolArgKeys>;
export type StringArgType = {
  [optionName in StringArgKeys]: IOptionFlag<string | undefined>;
};

export type FlagsParsed = {
  [Key in keyof AllOptions]: AllOptions[Key] extends boolean ? boolean : string | undefined;
};

export function toBooleanFlag(yargsBool: Options): IBooleanFlag<boolean> {
  const extendedDescription = `${yargsBool.describe} [default: ${yargsBool.default}].`;
  return flags.boolean({
    char: yargsBool.alias as AlphabetLowercase | AlphabetUppercase,
    description: extendedDescription,
    required: yargsBool.requiresArg,
    default: yargsBool.default,
    allowNo: yargsBool.default
  });
}

export function toStringFlag(yargsString: Options): IOptionFlag<string | undefined> {
  const options: Partial<IOptionFlag<string>> = {
    char: yargsString.alias as AlphabetLowercase | AlphabetUppercase,
    description: yargsString.describe,
    required: yargsString.requiresArg,
    default: yargsString.default
  };

  if (yargsString?.coerce?.name === 'resolve') {
    options.parse = (input: string, context: any): string => {
      let projectDir = '.';
      if (context.argv.includes('--projectDir')) {
        projectDir = context.argv[context.argv.indexOf('--projectDir') + 1];
      }
      return path.resolve(projectDir, input);
    };
  }
  return flags.string(options);
}
