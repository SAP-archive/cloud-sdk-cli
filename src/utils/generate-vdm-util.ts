/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */

import { flags } from '@oclif/command';
import { AlphabetLowercase, AlphabetUppercase } from '@oclif/parser/lib/alphabet';
import { IBooleanFlag, IOptionFlag } from '@oclif/parser/lib/flags';
import { GeneratorOptions as GeneratorOptionsSDK, generatorOptionsCli as generatorOptionsSDK } from '@sap/cloud-sdk-generator';
import * as path from 'path';
import { Options } from 'yargs';

interface GeneratorOptionCli {
  projectDir: Options;
}

export const generatorOptionCli: GeneratorOptionCli = {
  projectDir: {
    default: '.',
    describe: 'Path to the folder in which the VDM should be created. The input and output dir are relative to this directory.',
    type: 'string'
  }
};

type AllOptions = GeneratorOptionsSDK & GeneratorOptionCli;

// OClif distinguishes between boolean and string flags. Split the keys to get proper typing
type FilterBooleanKeys<Base> = {
  [Key in keyof Base]: Base[Key] extends boolean ? Key : never;
};

type BoolArgKeys = NonNullable<FilterBooleanKeys<AllOptions>[keyof AllOptions]>;
export type BoolArgType = {
  [optionName in BoolArgKeys]: IBooleanFlag<boolean>;
};

type StringArgKeys = keyof Omit<AllOptions, BoolArgKeys>;
export type StringArgType = {
  [optionName in StringArgKeys]: IOptionFlag<string | undefined>;
};

export type FlagsParsed = {
  [Key in keyof AllOptions]: AllOptions[Key] extends boolean ? boolean : string | undefined;
};

export function toGeneratorSDK(cliFlags: FlagsParsed): GeneratorOptionsSDK {
  return Object.entries(cliFlags)
    .filter(([key, value]) => typeof value !== 'undefined' && generatorOptionsSDK.hasOwnProperty(key))
    .reduce((prev, [key, value]) => ({ ...prev, [key]: value }), {}) as GeneratorOptionsSDK;
}

export function toBooleanFlag(yargsBool: Options): IBooleanFlag<boolean> {
  const extendedDescription = `${yargsBool.describe} [default: ${yargsBool.default}].`;
  return flags.boolean({
    char: yargsBool.alias as AlphabetLowercase | AlphabetUppercase,
    description: extendedDescription,
    required: yargsBool.requiresArg,
    default: yargsBool.default,
    allowNo: yargsBool.default
  });
}

export function toStringFlag(yargsString: Options): IOptionFlag<string | undefined> {
  const options: Partial<IOptionFlag<string>> = {
    char: yargsString.alias as AlphabetLowercase | AlphabetUppercase,
    description: yargsString.describe,
    required: yargsString.requiresArg,
    default: yargsString.default
  };

  if (yargsString?.coerce?.name === 'resolve') {
    options.parse = (input: string, context: any): string => {
      let projectDir = '.';
      if (context.argv.includes('--projectDir')) {
        projectDir = context.argv[context.argv.indexOf('--projectDir') + 1];
      }
      return path.resolve(projectDir, input);
    };
  }
  return flags.string(options);
}
