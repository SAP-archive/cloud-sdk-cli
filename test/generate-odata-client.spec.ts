/*!
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 */
import { GeneratorOptions as GeneratorOptionsSDK, generatorOptionsCli as generatorOptionsSDK } from '@sap/cloud-sdk-generator';
import * as fs from 'fs-extra';
import * as path from 'path';
import GenerateODataClient from '../src/commands/generate-odata-client';
import * as generateVdmUtil from '../src/utils/generate-odata-client-util';

const spyToGeneratorSDK = jest.spyOn(generateVdmUtil, 'toGeneratorSDK');

describe('generate-odata-client', () => {
  const pathForTests = path.resolve(__dirname, __filename.replace(/\./g, '-')).replace('-ts', '');

  beforeAll(() => {
    const pathForResources = path.resolve(__dirname, 'resources', 'template-generator-odata-client');
    fs.copySync(pathForResources, pathForTests);
  });

  afterAll(() => {
    fs.removeSync(pathForTests);
  });

  it('should fail if the mandatory parameters are not there', async () => {
    await expect(GenerateODataClient.run([])).rejects.toThrowError('-i, --inputDir INPUTDIR');
  });

  it('should parse strings and with --no- it should lead to all false.', async () => {
    const expected = getParsedInputWithAllBooleanFlagsFalse();
    delete expected.projectDir;

    await expect(GenerateODataClient.run([...getCliInputWithAllBooleanFlagsFalse(), '--projectDir', getProjectDir()])).rejects.toThrowError(
      'ENOENT: no such file or directory'
    );
    expect(spyToGeneratorSDK).toHaveReturnedWith(expected);
  });

  it('should resolve the projectdir', async () => {
    // TODO there was a issue in the SDK for the service-mapping.json file and windows. Onve the fix is out put this test back int.
    if (process.platform !== 'win32') {
      const expected = getDefault(getProjectDir());

      await expect(GenerateODataClient.run(['-i', 'input', '-o', 'output', '--projectDir', getProjectDir()])).rejects.toThrowError(
        'ENOENT: no such file or directory'
      );
      expect(spyToGeneratorSDK).toHaveReturnedWith(expected);
    }
  });

  function getProjectDir() {
    return path.resolve(__dirname, 'generate-odata-client-spec');
  }

  function getInputAndExpected(key: keyof GeneratorOptionsSDK): { expected: GeneratorOptionsSDK; args: string[] } | undefined {
    const args = ['-i', 'input', '-o', 'output', '--projectDir', getProjectDir()];
    const expected = getDefault(getProjectDir());

    const option = generatorOptionsSDK[key];
    if (option && option.type === 'boolean') {
      const casted = key as keyof GeneratorOptionsSDK;
      args.push(`--${option.default ? 'no-' : ''}${key}`);
      (expected[casted] as boolean) = !option.default;
      return { args, expected };
    }
  }

  it('should pass each boolean flags correctly', async () => {
    // TODO there was a issue in the SDK for the service-mapping.json file and windows. Onve the fix is out put this test back int.
    if (process.platform !== 'win32') {
      for (const key of Object.keys(generatorOptionsSDK)) {
        const argsExpected = getInputAndExpected(key as keyof GeneratorOptionsSDK);
        if (argsExpected) {
          await expect(GenerateODataClient.run(argsExpected.args)).rejects.toThrowError('ENOENT: no such file or directory');
          expect(spyToGeneratorSDK).toHaveReturnedWith(argsExpected.expected);
        }
      }
    }
  }, 10000);

  function getDefault(projectDir: string): GeneratorOptionsSDK {
    return {
      ...Object.keys(generatorOptionsSDK).reduce((prev, current) => {
        const value = generatorOptionsSDK[current as keyof GeneratorOptionsSDK];
        return value ? { ...prev, [current]: value.default } : prev;
      }, {} as any),
      inputDir: path.resolve(projectDir, 'input'),
      outputDir: path.resolve(projectDir, 'output'),
      serviceMapping: path.resolve(projectDir, 'input', 'service-mapping.json')
    };
  }

  function getCliInputWithAllBooleanFlagsFalse(): string[] {
    const allFalse = getParsedInputWithAllBooleanFlagsFalse();
    return Object.entries(allFalse).reduce((collected: string[], [key, value]) => {
      switch (typeof allFalse[key as keyof generateVdmUtil.FlagsParsed]) {
        case 'string':
          return [...collected, `--${key}`, value!.toString()];
        case 'boolean':
        default:
          return generatorOptionsSDK[key as keyof GeneratorOptionsSDK]?.default ? [...collected, `--no-${key}`] : collected;
      }
    }, []);
  }

  function getParsedInputWithAllBooleanFlagsFalse(): generateVdmUtil.FlagsParsed {
    return {
      aggregatorNpmPackageName: 'aggregatorNpm',
      aggregatorDirectoryName: 'aggregationDirectory',
      changelogFile: path.resolve(getProjectDir(), 'ChangeLogFile'),
      inputDir: path.resolve(getProjectDir(), 'InputDir'),
      outputDir: path.resolve(getProjectDir(), 'OutputDir'),
      projectDir: getProjectDir(),
      serviceMapping: path.resolve(getProjectDir(), 'ServiceMapping'),
      generateNpmrc: false,
      clearOutputDir: false,
      s4hanaCloud: false,
      sdkAfterVersionScript: false,
      writeReadme: false,
      useSwagger: false,
      generatePackageJson: false,
      generateTypedocJson: false,
      generateJs: false,
      generateCSN: false,
      forceOverwrite: false
    };
  }
});
