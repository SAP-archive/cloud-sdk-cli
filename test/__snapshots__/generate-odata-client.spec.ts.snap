// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-odata-client should fail if the mandatory parameters are not there 1`] = `
[Error: Missing required flag:
 -i, --inputDir INPUTDIR  [2mThis directory will be recursively searched for
                          .edmx/.xml files.[22m
See more help with --help]
`;
